flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/IcebergTableSink.java

```diff
@@ -0,0 +1,103 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.api.constraints.UniqueConstraint;
+import org.apache.flink.table.connector.ChangelogMode;
+import org.apache.flink.table.connector.sink.DataStreamSinkProvider;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+import org.apache.flink.table.connector.sink.abilities.SupportsOverwrite;
+import org.apache.flink.table.connector.sink.abilities.SupportsPartitioning;
+import org.apache.flink.types.RowKind;
+import org.apache.flink.util.Preconditions;
+import org.apache.iceberg.flink.sink.FlinkSink;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+
+public class IcebergTableSink implements DynamicTableSink, SupportsPartitioning, SupportsOverwrite {
+  private final TableLoader tableLoader;
+  private final TableSchema tableSchema;
+  private final ReadableConfig readableConfig;
+
+  private boolean overwrite = false;
+
+  private IcebergTableSink(IcebergTableSink toCopy) {
+    this.tableLoader = toCopy.tableLoader;
+    this.tableSchema = toCopy.tableSchema;
+    this.overwrite = toCopy.overwrite;
+    this.readableConfig = toCopy.readableConfig;
+  }
+
+  public IcebergTableSink(TableLoader tableLoader, TableSchema tableSchema, ReadableConfig readableConfig) {
+    this.tableLoader = tableLoader;
+    this.tableSchema = tableSchema;
+    this.readableConfig = readableConfig;
+  }
+
+  @Override
+  public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+    Preconditions.checkState(!overwrite || context.isBounded(),
+        "Unbounded data stream doesn't support overwrite operation.");
+
+    List<String> equalityColumns = tableSchema.getPrimaryKey()
+        .map(UniqueConstraint::getColumns)
+        .orElseGet(ImmutableList::of);
+
+    return (DataStreamSinkProvider) (providerContext, dataStream) -> FlinkSink.forRowData(dataStream)
+        .tableLoader(tableLoader)
+        .tableSchema(tableSchema)
+        .equalityFieldColumns(equalityColumns)
+        .overwrite(overwrite)
+        .flinkConf(readableConfig)
+        .append();
+  }
```

Comment:
Updated as the old signature didn't include the `providerContext` argument.

It's possible this change is causing the change data capture test to fail (possibly in the mapping to `UniqueConstraint`.